packages/flutter/lib/src/widgets/router.dart

```diff
@@ -663,11 +691,13 @@ class _RouterState<T> extends State<Router<T>> {
 
   Future<void> _rebuild([void value]) {
     setState(() {/* routerDelegate is ready to rebuild */});
+    _routeInformation.value = _retrieveNewRouteInformation();
```

Comment:
If we set the _routeInformation in _maybeNeedToReportRouteInformation, the only other places we may need to update the value will be in _processInitialRoute and _handleRouteInformationProviderNotification. In these two methods, we already know the routeInformation  because it is coming from the OS.
In some cases the Router may report a different route information after these two method,  this can happen if user want to redirect a route to a different route. but that it will be updated in _maybeNeedToReportRouteInformation when that happens.

Reply:
> In these two methods, we already know the routeInformation because it is coming from the OS.

Just because the OS told us to go to a place doesn't mean that we will go there. The RouterDelegate could reject the new location. Isn't `_rebuild` missing a call to `_maybeNeedToReportRouteInformation` to check whether we maybe need to tell the OS that we went to a different place instead?

Reply:
(I added that call at tests are all passing)